src/main/scala/com/databricks/spark/sql/perf/Tables.scala

```diff
@@ -211,9 +211,24 @@ abstract class Tables(sqlContext: SQLContext, scaleFactor: String,
           data.write
         }
       } else {
+        // treat non-partitioned tables as "one partition" that we want to coalesce
         if (clusterByPartitionColumns) {
-          // treat non-partitioned tables as "one partition" that we want to coalesce
-          data.coalesce(1).write
+          // in case data has more than maxRecordsPerFile, split into multiple writers to improve datagen speed
+          // files will be truncated to maxRecordsPerFile value, so the final result will be the same
+          val numRows = data.count
```

Comment:
Hmm... It's beneficial, but a bit ugly that this .count will cause the dsdgen / tpch datagen to be executed and do all the generating job twice.
Maybe for known scale factors you could hardcode a lookup table (tablename, sf) -> count, empty in Tables.scala, and overriden in TPCDS.scala and TPCH.scala?
Dirty few lines of code that could further shave a lot of time on high SF.